src/scene/Scene.hpp

```diff
@@ -56,12 +56,18 @@ struct Scene
 	void removeEntity(std::shared_ptr<Entity> entity);
 	void clear();
 
-	void setTime(Time time) { this->time = time; }
-	std::optional<Time> getTime() { return time; }
+	void setTime(Time time)
+	{
+		prevTime = this->time;
+		this->time = time;
+	}
+	std::optional<Time> getTime() const { return time; }
+	std::optional<Time> getPrevTime() const { return prevTime; }
+	std::optional<Time> getDeltaTime() const { return prevTime.has_value() ? std::optional(*time - *prevTime) : std::nullopt; }
 
-	std::size_t getObjectCount();
+	std::size_t getObjectCount() const;
 
-	CudaStream::Ptr getStream();
+	CudaStream::Ptr getStream() const;
```

Comment:
This method gives access to stream, that is a private member of Scene - consider not exposing it. Unless it is driven by other design choices obviously.

Reply:
I do not have better solution with std::shared_ptr - we would like to restrict access but also use a tool for sharing. In such case, you can e.g. return const ref to the object itself. However, I assume that this would not be convenient here, because of previous design choices.